docs/source/random_greedy_path.rst

```diff
@@ -0,0 +1,123 @@
+want to consider the ``'random-greedy'`` path optimizer. This samples many
+greedy paths and selects the best one found, which can often be exponentially
+better than the average.
+
+.. code:: python
+
+    import opt_einsum as oe
+    import numpy as np
+    import math
+
+    eq, shapes = oe.helpers.rand_equation(40, 5, seed=1, d_max=2)
+    arrays = list(map(np.ones, shapes))
+
+    path_greedy = oe.contract_path(eq, *arrays, optimize='greedy')[1]
+    print(math.log2(path_greedy.opt_cost))
+    # 36.04683022558587
+
+    path_rand_greedy = oe.contract_path(eq, *arrays, optimize='random-greedy')[1]
+    print(math.log2(path_rand_greedy.opt_cost))
+    # 32.203616699170865
+
+So here the random-greedy approach has found a path about
+16 times quicker (``= 2^(36 - 32)``).
+
+This approach works by randomly choosing from the best ``n`` contractions at
+each step, weighted by a
+`Boltzmann factor <https://en.wikipedia.org/wiki/Boltzmann_distribution>`_ with
+respect to the contraction with the 'best' cost. As such, contractions with
+very similar costs will be explored with equal probability, whereas those with
+higher costs will be less likely, but still possible. In this way, the
+optimizer can randomly explore the huge space of possible paths, but in a
+guided manner.
+
+The following graph roughly demonstrates the potential benefits of the
+``'random-greedy'`` algorithm, here for large randomly generated contractions,
+with either 8, 32 (the default), or 128 repeats:
+
+.. image:: path_found_flops_random.png
+    :width: 750
+
+.. note::
+
+    Bear in mind that such speed-ups are not guaranteed - it very much depends
+    on how structured or complex your contractions are.
+
+
+Customizing the Random-Greedy Path
+----------------------------------
+
+The random-greedy optimizer can be customized by instantiating your own
+:class:`~opt_einsum.path_random.RandomGreedy` object. Here you can control:
+
+* ``temperature`` - how far to stray from the locally 'best' contractions
+* ``rel_temperature`` - whether to normalize the temperature
+* ``nbranch`` - how many contractions (branches) to consider at each step
+* ``cost_fn`` - how to cost potential contractions
+
+There are also the main :class:`~opt_einsum.path_random.RandomOptimizer`
+options:
+
+* ``max_repeats`` - the maximum number of repeats
+* ``max_time`` - the maximum amount of time to run for (in seconds)
+* ``minimize`` - whether to minimize for total ``'flops'`` or ``'size'`` of the
+  largest intermediate
+
+For example, here we'll create an optimizer, then change its temperature
+whilst reusing it. We'll also set a high ``max_repeats`` and instead use a
+maximum time to terminate the search:
+
+.. code:: python
+
+    optimizer = oe.RandomGreedy(max_time=2, max_repeats=1_000_000)
+
+    for T in [1000, 100, 10, 1, 0.1]:
+        optimizer.temperature = T
+        path_rand_greedy = oe.contract_path(eq, *arrays, optimize=optimizer)[1]
```

Comment:
Could you potentially print out the best cost at each iteration? Could be interesting for demonstration.

Reply:
That would be nice - I'll add.